arbitrary_self_types: Split the Autoderef chain

dingxiangfei2009 opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/146095)*

This patch separate the two traits and isolate their use in the language better.

## Compiler perspective

This means that in method probing, we will chase down a `Deref` chain for an applicable type for the variable `self`, from which we additionally chase down a Receiver chain for an applicable `Self` type from a list of methods that accepts one of these possible types of `self`.

What matters is that whether enabling `arbitrary_self_types` or not ...
- we make sure that method resolution results stay the same under the respective modes, so the only expected changes to UI tests are only so that `impl Receiver` is now mandatory and the wording on the diagnostics is tweaked to a very small extent;
- the perf results should be comparable, even when `arbitrary_self_types` is enabled; the only perf impact would come from resolving methods on types with `Deref::Target <> Receiver::Targe` in conjunction with `#![feature(arbitrary_self_types_split_chain)]` because we need explore more types out of necessity.

### Changes to the inherent method candidate assembly

The inherent method candidate assembly collects the candidate `Self` and `self` types into `AutoderefCache` type, which maps one observed type to its `Deref::Target` and `Receiver::Target` types.

This is because that it is very common that `Deref::Target` coincides with `Receiver::Target`. It would present a performance impact if we do not cache the projections or obligations from these projections as soon as we stop assuming that `impl Deref` implies `impl Receiver`.

Another observation is that the current prober will still try to match the `self` after *n + 1* dereferences and call `xform_self_ty` to unify its type with some `Self` that is after *1, 2, ..., n*-th dereference. This part of the computation is wasteful.

## Language perspective

Receiver chain has been an extension of Deref chain, but the consequence has been questioned as this would admit method signatures that are deemed dubious.

It is also debatable that Receiver trait which determines applicable `Self` type should have anything to do with dereference operation in general.

## t-lang discussion

To facilitate discussion, we can use the Zulip thread [#t-lang > Consequences of making Deref a subtrait of Receiver](https://rust-lang.zulipchat.com/#narrow/channel/213817-t-lang/topic/Consequences.20of.20making.20Deref.20a.20subtrait.20of.20Receiver).

#### Pending items
- [x] First, gather feedback on the proposed change and assess the impact on the overall feature.
- [x] Stomp out UI test changes due to lost diagnostics, by enabling search through Receiver side-chains, so that we can suggest enabling `arbitrary_self_types` gate again.

rust-log-analyzer commented:
The job **`pr-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17382500825/job/49342995758) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49342995758) [(plain)](https://github.com/rust-lang/rust/commit/bef5ba5cf8593b131661c878fb2137cf83e3c90a/checks/49342995758/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] rustc_ast_lowering test:false 3.690
    Checking rustc_public v0.1.0-preview (/checkout/compiler/rustc_public)
[RUSTC-TIMING] rustc_metadata test:false 5.711
    Checking rustc_resolve v0.0.0 (/checkout/compiler/rustc_resolve)
error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
  --> compiler/rustc_resolve/src/ident.rs:48:19
   |
48 |         mut self: CmResolver<'r, 'ra, 'tcx>,
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or some type implementing `Receiver`
   = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/ident.rs:379:15
    |
379 |         self: CmResolver<'r, 'ra, 'tcx>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/ident.rs:740:15
    |
740 |         self: CmResolver<'r, 'ra, 'tcx>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/ident.rs:753:15
    |
753 |         self: CmResolver<'r, 'ra, 'tcx>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/ident.rs:794:19
    |
794 |         mut self: CmResolver<'r, 'ra, 'tcx>,
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/ident.rs:1069:19
     |
1069 |         mut self: CmResolver<'r, 'ra, 'tcx>,
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/ident.rs:1432:15
     |
1432 |         self: CmResolver<'r, 'ra, 'tcx>,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/ident.rs:1451:15
     |
1451 |         self: CmResolver<'r, 'ra, 'tcx>,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/ident.rs:1472:19
     |
1472 |         mut self: CmResolver<'r, 'ra, 'tcx>,
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/imports.rs:842:37
    |
842 |     fn resolve_import<'r>(mut self: CmResolver<'r, 'ra, 'tcx>, import: Import<'ra>) -> usize {
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/macros.rs:711:15
    |
711 |         self: CmResolver<'r, 'ra, 'tcx>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/macros.rs:734:19
    |
734 |         mut self: CmResolver<'r, 'ra, 'tcx>,
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/macros.rs:1095:19
     |
1095 |         mut self: CmResolver<'r, 'ra, 'tcx>,
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/lib.rs:1836:19
     |
1836 |         mut self: CmResolver<'r, 'ra, 'tcx>,
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/lib.rs:2232:19
     |
2232 |         mut self: CmResolver<'r, 'ra, 'tcx>,
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

[RUSTC-TIMING] rustc_public test:false 5.414
For more information about this error, try `rustc --explain E0307`.
[RUSTC-TIMING] rustc_resolve test:false 7.525
error: could not compile `rustc_resolve` (lib) due to 15 previous errors
```

</details>

rust-log-analyzer commented:
The job **`pr-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17386439388/job/49353639847) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49353639847) [(plain)](https://github.com/rust-lang/rust/commit/f3e8785c64f75f406717f05fb76b170d638fb043/checks/49353639847/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] rustc_metadata test:false 5.806
    Checking rustc_resolve v0.0.0 (/checkout/compiler/rustc_resolve)
[RUSTC-TIMING] rustc_monomorphize test:false 1.737
    Checking rustc_public v0.1.0-preview (/checkout/compiler/rustc_public)
error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
  --> compiler/rustc_resolve/src/ident.rs:48:19
   |
48 |         mut self: CmResolver<'r, 'ra, 'tcx>,
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or some type implementing `Receiver`
   = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/ident.rs:379:15
    |
379 |         self: CmResolver<'r, 'ra, 'tcx>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/ident.rs:740:15
    |
740 |         self: CmResolver<'r, 'ra, 'tcx>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/ident.rs:753:15
    |
753 |         self: CmResolver<'r, 'ra, 'tcx>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/ident.rs:794:19
    |
794 |         mut self: CmResolver<'r, 'ra, 'tcx>,
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/ident.rs:1069:19
     |
1069 |         mut self: CmResolver<'r, 'ra, 'tcx>,
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/ident.rs:1432:15
     |
1432 |         self: CmResolver<'r, 'ra, 'tcx>,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/ident.rs:1451:15
     |
1451 |         self: CmResolver<'r, 'ra, 'tcx>,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/ident.rs:1472:19
     |
1472 |         mut self: CmResolver<'r, 'ra, 'tcx>,
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/imports.rs:842:37
    |
842 |     fn resolve_import<'r>(mut self: CmResolver<'r, 'ra, 'tcx>, import: Import<'ra>) -> usize {
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/macros.rs:711:15
    |
711 |         self: CmResolver<'r, 'ra, 'tcx>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
   --> compiler/rustc_resolve/src/macros.rs:734:19
    |
734 |         mut self: CmResolver<'r, 'ra, 'tcx>,
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/macros.rs:1095:19
     |
1095 |         mut self: CmResolver<'r, 'ra, 'tcx>,
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/lib.rs:1836:19
     |
1836 |         mut self: CmResolver<'r, 'ra, 'tcx>,
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `RefOrMut<'r, Resolver<'ra, 'tcx>>`
    --> compiler/rustc_resolve/src/lib.rs:2232:19
     |
2232 |         mut self: CmResolver<'r, 'ra, 'tcx>,
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: type of `self` must be `Self` or some type implementing `Receiver`
     = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

[RUSTC-TIMING] rustc_public test:false 5.542
For more information about this error, try `rustc --explain E0307`.
[RUSTC-TIMING] rustc_resolve test:false 7.477
error: could not compile `rustc_resolve` (lib) due to 15 previous errors
```

</details>

rust-log-analyzer commented:
The job **`pr-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17386795118/job/49354530919) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49354530919) [(plain)](https://github.com/rust-lang/rust/commit/31788b7782503641c00dc9315febe3ce8081e71a/checks/49354530919/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking ui_test v0.30.2
error[E0307]: invalid `self` parameter type: `FileDescriptionRef<T>`
  --> src/tools/miri/src/shims/files.rs:99:26
   |
99 |     fn into_rc_any(self: FileDescriptionRef<Self>) -> Rc<dyn Any> {
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or some type implementing `Receiver`
   = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<T>`
   --> src/tools/miri/src/shims/files.rs:104:15
    |
104 |         self: FileDescriptionRef<Self>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<Stdin>`
   --> src/tools/miri/src/shims/files.rs:232:15
    |
232 |         self: FileDescriptionRef<Self>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<std::io::Stdout>`
   --> src/tools/miri/src/shims/files.rs:259:15
    |
259 |         self: FileDescriptionRef<Self>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<Stderr>`
   --> src/tools/miri/src/shims/files.rs:289:15
    |
289 |         self: FileDescriptionRef<Self>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<FileHandle>`
   --> src/tools/miri/src/shims/files.rs:319:15
    |
319 |         self: FileDescriptionRef<Self>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<FileHandle>`
   --> src/tools/miri/src/shims/files.rs:333:15
    |
333 |         self: FileDescriptionRef<Self>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<NullOutput>`
   --> src/tools/miri/src/shims/files.rs:417:15
    |
417 |         self: FileDescriptionRef<Self>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<AnonSocket>`
   --> src/tools/miri/src/shims/unix/unnamed_socket.rs:105:15
    |
105 |         self: FileDescriptionRef<Self>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<AnonSocket>`
   --> src/tools/miri/src/shims/unix/unnamed_socket.rs:116:15
    |
116 |         self: FileDescriptionRef<Self>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<EventFd>`
  --> src/tools/miri/src/shims/unix/linux_like/eventfd.rs:55:15
   |
55 |         self: FileDescriptionRef<Self>,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or some type implementing `Receiver`
   = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<EventFd>`
  --> src/tools/miri/src/shims/unix/linux_like/eventfd.rs:88:15
   |
88 |         self: FileDescriptionRef<Self>,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or some type implementing `Receiver`
   = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<Self>`
  --> src/tools/miri/src/shims/files.rs:87:26
   |
87 |     fn into_rc_any(self: FileDescriptionRef<Self>) -> Rc<dyn Any>;
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or some type implementing `Receiver`
   = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<Self>`
  --> src/tools/miri/src/shims/files.rs:92:15
   |
92 |         self: FileDescriptionRef<Self>,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or some type implementing `Receiver`
   = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<Self>`
   --> src/tools/miri/src/shims/files.rs:143:15
    |
143 |         self: FileDescriptionRef<Self>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error[E0307]: invalid `self` parameter type: `FileDescriptionRef<Self>`
   --> src/tools/miri/src/shims/files.rs:160:15
    |
160 |         self: FileDescriptionRef<Self>,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: type of `self` must be `Self` or some type implementing `Receiver`
    = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

[RUSTC-TIMING] ui_test test:false 1.633
For more information about this error, try `rustc --explain E0307`.
[RUSTC-TIMING] miri test:false 9.969
error: could not compile `miri` (lib) due to 16 previous errors
```

</details>

dingxiangfei2009 commented:
@rustbot label +F-arbitrary_self_types

dingxiangfei2009 committed:
arbitrary_self_types: split the Autoderef chain

Receiver chain has been an extension of Deref chain,
but the consequence has been questioned as this would admit method
signatures that are deemed dubious.

It is also debatable that Receiver trait which determines applicable
`Self` type should have anything to do with dereference operation
in general.

This patch separate the two traits and isolate their use in the language.
This means that in method probing, we will chase down a Deref chain for
an applicable type for the variable `self`, from which we additionally
chase down a Receiver chain for an applicable `Self` type from a list of
methods that accepts one of these possible types of `self`.

This patch includes rewording of diagnostics.

jackh726 commented:
Is this waiting on discussion and a decision from the lang team?

bors commented:
:umbrella: The latest upstream changes (presumably #145993) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dingxiangfei2009 commented:
The report is a bit overdue. I would like to just publish a partial text while I need more time to restore some diagnostics.

## Acceptable receivers

Landing #146095 means that types implementing `Deref` are not automatically acceptable receiver types anymore.

<table>
    <tr>
        <th>Before #146095</th>
        <th>After #146095</th>
    </tr>
    <tr>
<td>

```rust
struct A;
struct B;
impl Deref for A {
    type Target = B;
    fn deref(&self) -> &B {
        &B
    }
}
impl B {
    fn recv(self: &A);
}

A.recv(); // resolves to B::recv
```

</td>
<td>

```rust
struct A;
struct B;
impl Deref for A {
    type Target = B;
    fn deref(&self) -> &B {
        &B
    }
}
// This is needed ...
impl Receiver for A {
    type Target = B;
}
impl B {
    fn recv(self: &A);
}

A.recv();
// ^ ... in order to resolve this to B::recv
```

</td>
</tr>
    
<tr>
<td>

```rust
// The following passes checking in our surprises
impl Trait for str {
    fn recv(self: &String) {}
}
"hello".to_string().recv();
impl Trait for MyType {
    fn recv(self: MutexGuard<'_, Self>) {}
}
let my_mutex = Mutex::new(MyType);
my_mutex.lock().recv();
```

</td>
    
<td>
    
```rust
// The following stops to compile for good reasons
impl Trait for str {
    fn recv(self: &String) {}
    // `String: Deref<Target = str>` but without `Receiver<Target = str>`
}
"hello".to_string().recv(); //~ ERROR
impl Trait for MyType {
    fn recv(self: MutexGuard<'_, Self>) {}
    // `MutexGuard<'_, MyType>: Deref<Target = MyType>` but without `Receiver<Target = MyType>`
}
let my_mutex = Mutex::new(MyType);
my_mutex.lock().recv(); //~ ERROR
```
</td>
</tr>
</table>

    
    
## Library changes
This patch demonstrates that we now explicitly mark types like `Rc`, `Arc`, `Box`, `Pin` as valid receivers regardless of whether a `Deref` is applicable.
    
This is particularly important for `Pin` as we want to conditionally make it `Deref`able but unconditionally a valid method receiver just like the stable Rust.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18263583815/job/51994819167) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51994819167) [(plain)](https://github.com/rust-lang/rust/commit/423792b4e601f7015d257949d52b4453e96e9db9/checks/51994819167/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0405`.
"/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/dist/rustc-clif" "-Csymbol-mangling-version=v0" "-Zrandomize-layout" "-Zunstable-options" "--check-cfg=cfg(bootstrap)" "-Zmacro-backtrace" "-Csplit-debuginfo=off" "-Clink-arg=-L/usr/lib/llvm-20/lib" "-Cllvm-args=-import-instr-limit=10" "-Clink-args=-Wl,-z,origin" "-Clink-args=-Wl,-rpath,$ORIGIN/../lib" "-Alinker-messages" "-L" "crate=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/build/example" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/build/example" "-Cdebuginfo=2" "--target" "aarch64-unknown-linux-gnu" "-Cpanic=abort" "--check-cfg=cfg(jit)" "example/arbitrary_self_types_pointers_and_wrappers.rs" exited with status ExitStatus(unix_wait_status(256))
Bootstrap failed while executing `--stage 2 test --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Command `/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo run --target aarch64-unknown-linux-gnu -Zbinary-dep-depinfo -j 4 -Zroot-dir=/checkout --locked --color always --release --manifest-path /checkout/compiler/rustc_codegen_cranelift/build_system/Cargo.toml -- test --download-dir /checkout/obj/build/cg_clif_download --out-dir /checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif --no-unstable-features --use-backend cranelift --sysroot llvm --skip-test testsuite.extended_sysroot [workdir=/checkout/compiler/rustc_codegen_cranelift]` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/test.rs:3658:25
Executed at: src/bootstrap/src/core/build_steps/test.rs:3703:26

Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:24:56
  local time: Sun Oct  5 20:31:44 UTC 2025
  network time: Sun, 05 Oct 2025 20:31:45 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18264057418/job/51995919922) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51995919922) [(plain)](https://github.com/rust-lang/rust/commit/962a73764419e8a6b2e9a53390650b3e90cf5f4d/checks/51995919922/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- library/core/src/marker.rs - marker::CoercePointee (line 1192) stdout ----
error[E0307]: invalid `self` parameter type: `MySmartPointer<Self>`
  --> library/core/src/marker.rs:1211:19
   |
22 |     fn func(self: MySmartPointer<Self>);
   |                   ^^^^^^^^^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or some type implementing `Receiver`
   = help: consider changing to `self`, `&self`, `&mut self`, or a type implementing `Receiver` such as `self: Box<Self>`, `self: Rc<Self>`, or `self: Arc<Self>`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0307`.
Couldn't compile the test.
```

</details>

jackh726 commented on compiler/rustc_hir_analysis/messages.ftl:
```diff
@@ -264,7 +264,8 @@ hir_analysis_invalid_receiver_ty_help =
 
 hir_analysis_invalid_receiver_ty_help_no_arbitrary_self_types =
-    consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)
+    consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`);
```
I'm surprised by this change. I would expect to only change the diagnostic when the feature is not enabled `hir_analysis_invalid_receiver_ty_help`

jackh726 commented on compiler/rustc_hir_typeck/src/method/confirm.rs:
```diff
@@ -350,18 +351,37 @@ impl<'a, 'tcx> ConfirmContext<'a, 'tcx> {
+                    }
+                }
+                if follow_receiver_chain {
+                    EitherIter::A(fcx.autoderef(span, ty).follow_receiver_chain(), PhantomData)
```
Okay, so if I'm understanding this correctly, for *each* step in `Deref` chain, we then try a `Receiver` chain? That's pretty crazy, and is very likely going to have an effect on perf.

I would really like to evaluate perf before and after this PR (when enabling arbitrary self types).

dingxiangfei2009 replied:
I think we are already performing quadratic time probing with how `Receiver` is set up on `main` branch.

- `FnCtxt::probe_op` walks the `Deref` chain followed by `Receiver` extension chain for candidate `Self` type. They are collected into `steps`.
- `ProbeContext::assemble_*` collects all methods with applicable `Self` type, for each `Self` from `steps`.
- `ProbeContext::pick_all_method` does the following. For each type from `steps`, *only when it is a member of the `Deref` chain which corresponds to a possible `self` value*, this function calls out to `ProbeContext::pick_method` among *all* candidate method items. It will try to unify the type of `self` with that of `self` argument in the method signature. That is how we find a method match with one `self` and one `Self` in a right `impl`.

I am thinking of sorting the candidates into bins for each candidate `Self` type. With this structure we save some search effort. Let me work on a patch on this idea.

jackh726 reviewed (commented):
Wasn't sure if this was quite ready for review or not. But, here are my thoughts.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146095/bd3487101ff9397ca8a54dd83aa360d93342b82e..32433c2d288e77d188cc9fe734268c62e50fc7d1)*

bors commented:
:umbrella: The latest upstream changes (presumably #144420) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dingxiangfei2009 committed:
Enable experimentation with split/merged chains

To allow use of Receiver with Target diverging from Deref::Target,
one must now enable an unstable feature flag arbitrary-self-types-split-chains.

Signed-off-by: Xiangfei Ding <dingxiangfei2009@protonmail.ch>

dingxiangfei2009 committed:
Pick methods from the right receivers

Previously we have been picking methods for receiver types even though
we are inspecting other receiver types that are several levels of
dereferences, which is wasted effort because there will never be matches
with these methods.

Signed-off-by: Xiangfei Ding <dingxiangfei2009@protonmail.ch>

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued 7757954de1446aa7950e3bf5df62e7df2c06fac5 with parent 4c62aa446803c8f31eafdc0c0119209ff129b3e0, future [comparison URL](https://perf.rust-lang.org/compare.html?start=4c62aa446803c8f31eafdc0c0119209ff129b3e0&end=7757954de1446aa7950e3bf5df62e7df2c06fac5).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.0 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (7757954de1446aa7950e3bf5df62e7df2c06fac5): [comparison URL](https://perf.rust-lang.org/compare.html?start=4c62aa446803c8f31eafdc0c0119209ff129b3e0&end=7757954de1446aa7950e3bf5df62e7df2c06fac5&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=4c62aa446803c8f31eafdc0c0119209ff129b3e0&end=7757954de1446aa7950e3bf5df62e7df2c06fac5&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.1%, 0.3%]   | 18    |
| Regressions ❌ <br /> (secondary)  | 0.4%  | [0.1%, 0.7%]   | 10    |
| Improvements ✅ <br /> (primary)   | -0.5% | [-1.2%, -0.1%] | 16    |
| Improvements ✅ <br /> (secondary) | -0.1% | [-0.3%, -0.1%] | 10    |
| All ❌✅ (primary)                 | -0.1% | [-1.2%, 0.3%]  | 34    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4c62aa446803c8f31eafdc0c0119209ff129b3e0&end=7757954de1446aa7950e3bf5df62e7df2c06fac5&stat=max-rss)
<details>
<summary>Results (primary 0.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.6%  | [1.5%, 1.8%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.6% | [-2.2%, -0.9%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.0%  | [-2.2%, 1.8%]  | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4c62aa446803c8f31eafdc0c0119209ff129b3e0&end=7757954de1446aa7950e3bf5df62e7df2c06fac5&stat=cycles:u)
<details>
<summary>Results (secondary -2.5%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.5% | [-3.2%, -2.0%] | 5     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=4c62aa446803c8f31eafdc0c0119209ff129b3e0&end=7757954de1446aa7950e3bf5df62e7df2c06fac5&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 476.46s -> 476.713s (0.05%)
**Artifact size**: 390.83 MiB -> 390.87 MiB (0.01%)

dingxiangfei2009 commented:
@rustbot label: +perf-regression-triaged

jackh726 commented on compiler/rustc_hir_analysis/src/coherence/builtin.rs:
```diff
@@ -792,3 +794,55 @@ fn visit_implementation_of_coerce_pointee_validity(
+            tcx.sess,
+            sym::arbitrary_self_types_split_chains,
+            span,
+            "`Receiver::Target` diverging from `Deref::Target` is not supported; we look forward to your feedback",
```
This should link an issue.

jackh726 reviewed (commented):
Not sure where this is in terms of being ready for review, but took another pass at looking this over.

First, as a general thought: I have pretty much no clue what the expected behavior here is supposed to be, if this has been discussed among the lang team, etc.... I think it makes sense to have some documentation either in-tree or in the rustc-dev-guide that explains the high-level details for the relationship between `Deref` and `Receiver`. For example, I looked at `receiver-complex.rs` and my eyes went wide because I had basically no clue what intuition was for the behavior.

Second, I feel like there's a lot happening in this PR. I'm wondering if there are things we can split out to a separate PR, or things that we can split into separate commits. One easy thing would be to make all added tests be in their own commit with the behavior on main - and then have each commit update. Other than that, I'm really not sure.

Third, I haven't quite read through the last commit. The changes the `method/probe.rs` are *rough*, and I personally don't want to land them as-is; they definitely seem like a maintenance hazard, and I'm sure the rest of the types team would agree. Perhaps I need to read through the changes to try to be able to make some suggestions for alternatives, but right now I don't have any.

Fourth: comments, comments, comments. There's a lot of code added and changed in this PR and very few comments explaining what the code does or why.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146095/3b1b0ef4d80d3117924d91352c8b6ca528708b3c..be635d1de6fbc6d45aeacbccc1866cbf11a48b20)*

jackh726 commented:
I am going to change this to S-blocked and S-waiting-on-author. There's an open zulip topic now in https://rust-lang.zulipchat.com/#narrow/channel/144729-t-types/topic/.60arbitrary_self_types.60.3A.20Split.20the.20Autoderef.20chain.20.28.23140095.29/with/583211400

In general, while I *could* do some more review on this PR directly, I think what is really needed is better (or at least more available) docs or a design doc for this. So I can review this with some context.
